src/main_alt.ts

```diff
@@ -97,27 +98,72 @@ function writeUnionType(w: CodeBlockWriter, node: UnionType) {
   w.writeLine('rt.Union(');
   for (const type of node.types) {
     writeAnyType(w, type);
-    w.write(', ');
+    w.write(',\n');
   }
-  w.write(') ');
+  w.writeLine(') ');
 }
 
 function writeIntersectionType(w: CodeBlockWriter, node: UnionType) {
   w.writeLine('rt.Intersect(');
   for (const type of node.types) {
     writeAnyType(w, type);
-    w.write(', ');
+    w.write(',\n');
   }
-  w.write(') ');
+  w.writeLine(') ');
+}
+
+/**
+ * public for testing
+ * @private
+ * @param fields
+ */
+export function groupFieldKinds(
+  fields: readonly RecordField[],
+): {
+  readonly: boolean;
+  nullable: boolean;
+  fields: RecordField[];
+}[] {
+  return [
+    {
+      readonly: false,
+      nullable: false,
+      fields: fields.filter((e) => !e.readonly && !e.nullable),
+    },
+    {
+      readonly: false,
+      nullable: true,
+      fields: fields.filter((e) => !e.readonly && e.nullable),
+    },
+    {
+      readonly: true,
+      nullable: false,
+      fields: fields.filter((e) => e.readonly && !e.nullable),
+    },
+    {
+      readonly: true,
+      nullable: true,
+      fields: fields.filter((e) => e.readonly && e.nullable),
+    },
+  ].filter((e) => e.fields.length > 0);
 }
 
 function writeRecordType(w: CodeBlockWriter, node: RecordType) {
-  w.writeLine('rt.Record({');
-  for (const field of node.fields) {
-    w.write(field.name);
-    w.write(': ');
-    writeAnyType(w, field.type);
-    w.write(',');
+  const fieldKinds = groupFieldKinds(node.fields);
+  const hasMultiple = fieldKinds.length > 1;
+  w.conditionalWriteLine(hasMultiple, 'rt.intersect(');
+  for (const fieldKind of fieldKinds) {
+    w.writeLine('rt.Record({');
+    for (const field of fieldKind.fields) {
+      w.write(field.name);
+      w.write(': ');
+      writeAnyType(w, field.type);
+      w.write(',\n');
+    }
+    w.write('})');
+    w.conditionalWrite(fieldKind.nullable ?? false, '.asPartial()');
+    w.conditionalWrite(fieldKind.readonly ?? false, '.asReadonly()');
```

Comment:
Nice :)

So `groupFieldKinds` is used for these conditions. They are evaluated based on the `node.fields` and if there are `readonly` and/or `nullable` fields.

Reply:
Yup. No other way, right now, to make a `Record` where some stuff is partial and some is nullable. You can use `.And(` but that's just an intersection with different syntax. There's new stuff that just landed in runtypes that can  be used as well when it's released: https://github.com/pelotom/runtypes#optional-values